src/puppet.ts

```diff
@@ -378,8 +362,71 @@ export abstract class Puppet extends PuppetEventEmitter {
    *
    *
    */
-  public abstract async start () : Promise<void>
-  public abstract async stop ()  : Promise<void>
+  public async start () : Promise<void> {
+    this.on('heartbeat', this.feedDog)
+    this.watchdog.on('reset', this.dogReset)
+
+    /**
+     * 2. Setup `reset` Event via a 1 second Throttle Queue:
+     */
+    this.resetThrottleQueue = new ThrottleQueue<string>(1000)
+    // 2.2. handle all `reset` events via the resetThrottleQueue
+    this.on('reset', this.throttleReset)
+    // 2.3. call reset() and then ignore the following `reset` event for 1 second
+    this.resetThrottleQueueSubscription = this.resetThrottleQueue.subscribe(reason => {
+      log.silly('Puppet', 'constructor() resetThrottleQueue.subscribe() reason: "%s"', reason)
+      this.reset(reason)
+    })
+  }
+
+  public async stop (): Promise<void> {
+    this.removeListener('heartbeat', this.feedDog)
+    this.watchdog.removeListener('reset', this.dogReset)
+    this.removeListener('reset', this.throttleReset)
+
+    if (this.resetThrottleQueueSubscription) {
+      this.resetThrottleQueueSubscription.unsubscribe()
+      this.resetThrottleQueueSubscription = undefined
+    }
+    if (this.resetThrottleQueue) {
+      this.resetThrottleQueue.unsubscribe()
+      this.resetThrottleQueue = undefined
+    }
+
+    this.watchdog.sleep()
+
+    /**
+     * FIXME: clear cache when stop
+     * keep the cache as a temp workaround since wechaty-puppet-hostie has reconnect issue
+     * with un-cleared cache in wechaty-puppet will make the reconnect recoverable
+     *
+     * Related issue: https://github.com/wechaty/wechaty-puppet-hostie/issues/31
+     */
+    // this.cacheContactPayload.clear()
+    // this.cacheFriendshipPayload.clear()
+    // this.cacheMessagePayload.clear()
+    // this.cacheRoomPayload.clear()
+    // this.cacheRoomInvitationPayload.clear()
+    // this.cacheRoomMemberPayload.clear()
+  }
+
+  private feedDog (payload: any) {
+    this.watchdog.feed(payload)
+  }
+
+  private dogReset (lastFood: any) {
```

Comment:
Do those three private functions a must for existing? It seems that they only are used once and are all very short, so I'd like to suggest that those codes can be inlined directly for not increasing our methods in the class for future maintains easier.

Reply:
The reason I added those functions is that I want to fully recover the status of the puppet before the `start()` call when we call the `stop()` on it. And I only operates the one that we added in the `wechaty-puppet` class, so I need to keep the reference to the listeners to only clear it in the `stop()`.

Reply:
Could you please explain it more?

I'm afraid that I did not follow your points.

Can you archive your goals without creating those private methods? If not, why?

Reply:
Here are the code I use to remove the listener:
```ts
this.removeListener('heartbeat', this.feedDog)
```

I want to only remove the listener that I added in the `wechaty-puppet` class, without touch any other listeners that attached on the `wechaty-puppet`. So I need to keep the reference to the listener functions.

I can not use `removeAllListeners` here since the `wechaty-puppet` will be attached multiple listeners in `wechaty`, doing that will definitely break the logic in `wechaty`. From this perspective, I kept all the functions as private method in the class.

Do you know any other solutions to this problem?

Reply:
Thanks for the explanation, it's acceptable.

We can use your current design with this PR. For future solutions, I'll try to finger it out by RxJS, if there's any.